src/credentials/StaticGroupDataProvider.cpp

```diff
@@ -0,0 +1,514 @@
+        Fabric * fabric = GetFabric(fabric_index);
+        // Remove all mappings from fabric
+        for (uint16_t i = 0; fabric && i < kEndpointEntriesMax; ++i)
+        {
+            fabric->endpoints[i].in_use = false;
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    GroupMappingIterator * IterateGroupMappings(chip::FabricIndex fabric_index, EndpointId endpoint) override
+    {
+        return new EndpointIterator(GetFabric(fabric_index), endpoint);
+    }
+
+    // States
+    CHIP_ERROR SetGroupState(chip::FabricIndex fabric_index, GroupState & state) override
+    {
+        Fabric * fabric    = GetFabric(fabric_index);
+        StateEntry * entry = nullptr;
+
+        // Search for existing, or unused entry
+        for (uint16_t i = 0; fabric && i < kStateEntriesMax; ++i)
+        {
+            if (fabric->states[i].in_use)
+            {
+                if (fabric->states[i].state_index == state.state_index)
+                {
+                    // Reuse existing entry
+                    entry = &fabric->states[i];
+                    break;
+                }
+            }
+            else if (!entry)
+            {
+                // Unused entry
+                entry = &fabric->states[i];
+            }
+        }
+        if (entry)
+        {
+            entry->state_index   = state.state_index;
+            entry->group         = state.group;
+            entry->key_set_index = state.key_set_index;
+            entry->in_use        = true;
+            return CHIP_NO_ERROR;
+        }
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    CHIP_ERROR GetGroupState(chip::FabricIndex fabric_index, GroupState & state) override
+    {
+        Fabric * fabric = GetFabric(fabric_index);
+        // Search for existing state
+        for (uint16_t i = 0; fabric && i < kStateEntriesMax; ++i)
+        {
+            StateEntry & entry = fabric->states[i];
+            if (entry.in_use && entry.state_index == state.state_index)
+            {
+                // Found
+                state.group         = entry.group;
+                state.key_set_index = entry.key_set_index;
+                return CHIP_NO_ERROR;
+            }
+        }
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+
+    CHIP_ERROR RemoveGroupState(chip::FabricIndex fabric_index, uint16_t state_index) override
+    {
+        Fabric * fabric = GetFabric(fabric_index);
+        // Search for existing state
+        for (uint16_t i = 0; fabric && i < kStateEntriesMax; ++i)
+        {
+            StateEntry & entry = fabric->states[i];
+            if (entry.in_use && entry.state_index == state_index)
+            {
+                // Found
+                entry.in_use = false;
+                return CHIP_NO_ERROR;
+            }
+        }
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+
+    GroupStateIterator * IterateGroupStates(chip::FabricIndex fabric_index) override
+    {
+        return new StateIterator(GetFabric(fabric_index));
```

Comment:
Since this iterator is an imutable iterator, we should return it *by value* and have it handle its own memory management. Documentation can clearly state that only one can be used at a time. Same for all other iterators.